tests/devices/test_default_qutrit.py

```diff
@@ -0,0 +1,570 @@
+        (
+            3,
+            np.kron(np.eye(3), U_tadd) @ np.kron(np.eye(3), np.kron(U_thadamard_01, np.eye(3))),
+            three_wire_final_state / np.sqrt(2),
+        ),
+        (
+            4,
+            np.kron(U_tadd, U_tswap)
+            @ np.kron(U_thadamard_01, np.eye(27))
+            @ np.kron(np.eye(9), np.kron(U_thadamard_01, np.eye(3))),
+            four_wire_final_state / 2.0,
+        ),
+    ]
+
+    @pytest.mark.parametrize("num_wires, mat, expected_out", state_measurement_data)
+    def test_qutrit_circuit_state_measurement(self, num_wires, mat, expected_out, tol):
+        """Tests if state returned by state function is correct"""
+        dev = qml.device("default.qutrit", wires=num_wires)
+
+        @qml.qnode(dev)
+        def circuit(mat):
+            qml.QutritUnitary(mat, wires=list(range(num_wires)))
+            return qml.state()
+
+        state = circuit(mat)
+        assert np.allclose(state, expected_out, atol=tol)
+
+
+class TestProbabilityIntegration:
+    """Test probability method for when computation is/is not analytic"""
+
+    def mock_analytic_counter(self, wires=None):
+        self.analytic_counter += 1
+        return np.array([1, 0, 0, 0, 0, 0, 0, 0, 0], dtype=float)
+
+    @pytest.mark.parametrize(
+        "x", [[U_thadamard_01, U_clock], [U_shift, U_thadamard_01], [U_z_12, U_thadamard_01]]
+    )
+    def test_probability(self, x, tol):
+        """Test that the probability function works for finite and infinite shots"""
+        dev = qml.device("default.qutrit", wires=2, shots=1000)
+        dev_analytic = qml.device("default.qutrit", wires=2, shots=None)
+
+        def circuit(x):
+            qml.QutritUnitary(x[0], wires=0)
+            qml.QutritUnitary(x[1], wires=0)
+            qml.QutritUnitary(U_tadd, wires=[0, 1])
+            return qml.probs(wires=[0, 1])
+
+        prob = qml.QNode(circuit, dev)
+        prob_analytic = qml.QNode(circuit, dev_analytic)
+
+        assert np.isclose(prob(x).sum(), 1, atol=tol, rtol=0)
+        assert np.allclose(prob_analytic(x), prob(x), atol=0.1, rtol=0)
+        assert not np.array_equal(prob_analytic(x), prob(x))
+
+    def test_call_generate_samples(self, monkeypatch):
+        """Test analytic_probability call when generating samples"""
+        self.analytic_counter = False
+
+        dev = qml.device("default.qutrit", wires=2, shots=1000)
+        monkeypatch.setattr(dev, "analytic_probability", self.mock_analytic_counter)
+
+        # generate samples through `generate_samples` (using 'analytic_probability')
+        dev.generate_samples()
+
+        # should call `analytic_probability` once through `generate_samples`
+        assert self.analytic_counter == 1
+
+    def test_stateless_analytic_return(self):
+        """Test that analytic_probability returns None if device is stateless"""
+        dev = qml.device("default.qutrit", wires=2)
+        dev._state = None
+
+        assert dev.analytic_probability() is None
```

Comment:
Does this ever happen for `"default.qutrit" in practice?

Reply:
Not unless a curious user sets `dev._state` to `None` as is done in this test.